static/app/views/performance/content.tsx

```diff
@@ -33,86 +33,97 @@ import Onboarding from './onboarding';
 import {addRoutePerformanceContext, handleTrendsClick} from './utils';
 
 type Props = {
-  api: Client;
-  organization: Organization;
   selection: GlobalSelection;
   location: Location;
   router: InjectedRouter;
-  projects: Project[];
-  loadingProjects: boolean;
   demoMode?: boolean;
 };
 
 type State = {
   eventView: EventView;
-  error: string | undefined;
+  error?: string;
 };
 
-class PerformanceContent extends Component<Props, State> {
-  static getDerivedStateFromProps(nextProps: Readonly<Props>, prevState: State): State {
-    return {
-      ...prevState,
-      eventView: generatePerformanceEventView(
-        nextProps.organization,
-        nextProps.location,
-        nextProps.projects
-      ),
-    };
-  }
+function PerformanceContent({selection, location, demoMode}: Props) {
+  const api = useApi();
+  const organization = useOrganization();
+  const {projects} = useProjects();
+
+  const prevDeeplyNestedSelectProjects = usePrevious(selection.projects);
+  const prevDeeplyNestedSelectDateTime = usePrevious(selection.datetime);
 
-  state: State = {
-    eventView: generatePerformanceEventView(
-      this.props.organization,
-      this.props.location,
-      this.props.projects
-    ),
+  const [state, setState] = useState<State>({
+    eventView: generatePerformanceEventView(organization, location, projects),
     error: undefined,
-  };
+  });
 
-  componentDidMount() {
-    const {api, organization, selection} = this.props;
+  useEffect(() => {
     loadOrganizationTags(api, organization.slug, selection);
     addRoutePerformanceContext(selection);
     trackAnalyticsEvent({
       eventKey: 'performance_views.overview.view',
       eventName: 'Performance Views: Transaction overview view',
       organization_id: parseInt(organization.id, 10),
-      show_onboarding: this.shouldShowOnboarding(),
+      show_onboarding: shouldShowOnboarding(),
     });
-  }
+  }, []);
+
+  useEffect(() => {
+    setState({
+      ...state,
+      eventView: generatePerformanceEventView(organization, location, projects),
+    });
+  }, [organization, location, projects]);
 
-  componentDidUpdate(prevProps: Props) {
-    const {api, organization, selection} = this.props;
+  useEffect(() => {
     if (
-      !isEqual(prevProps.selection.projects, selection.projects) ||
-      !isEqual(prevProps.selection.datetime, selection.datetime)
+      !isEqual(prevDeeplyNestedSelectProjects, selection.projects) ||
+      !isEqual(prevDeeplyNestedSelectDateTime, selection.datetime)
     ) {
       loadOrganizationTags(api, organization.slug, selection);
       addRoutePerformanceContext(selection);
     }
-  }
+  }, [
+    prevDeeplyNestedSelectProjects,
```

Comment:
I don't think ref values have any meaning inside a useEffect dependency list?

Reply:
if prevDeeplyNestedSelectProjects gets updated, then it has an effect here, because it's a new value

Reply:
I'm pretty sure it's not idiomatic because `useRef` doesn't trigger a render unlike state. 

I made a sandbox to show that it doesn't work if nothing else is causing a render.
https://codesandbox.io/s/silent-sky-vjor2?file=/src/App.js

Reply:
I've previously also worked on this sandbox and everything seems to work as we expect https://codesandbox.io/s/useprevious-2hrj5?from-embed=&file=/src/App.jsx

Reply:
It works but it's an anti-pattern.

Reply:
sorry...wrong link https://codesandbox.io/s/useprevious-forked-fq4ue

Reply:
why do you think that? it seems like that it's a recommended pattern by the react community https://reactjs.org/docs/hooks-faq.html#how-to-get-the-previous-props-or-state

Reply:
Ah, sorry, I don't mean the `usePrevious` hook directly, it's returning a `ref.current`. I meant using what is essentially a `ref.current` in the dependency array of `useEffect` is recommended against because ref changes don't trigger a re-render, which means that `useEffect` won't even be called if it solely were to change. In the case of previous, it's matched with a current state, which is what is causing the re-render. The state is then probably sufficient in the dependency array of the `useEffect`. There is a react hook lint rule for this, but I'm guessing the custom hook hides it.

![Screen Shot 2021-11-24 at 11 25 03 AM](https://user-images.githubusercontent.com/6111995/143279275-e84dd21f-48be-4d3e-85ec-f0979be9b950.png)